contracts/lib/MerklePatriciaProof.sol

```diff
@@ -0,0 +1,255 @@
+/**
+ * @title MerklePatriciaVerifier
+ * @author Sam Mayo (sammayo888@gmail.com)
+ *
+ * @dev Library for verifing merkle patricia proofs.
+ */
+
+import "./RLP.sol";
+
+library MerklePatriciaProof {
+    /**
+     * @dev Verifies a merkle patricia proof.
+     * @param value The terminating value in the trie.
+     * @param encodedPath The path in the trie leading to value.
+     * @param rlpParentNodes The rlp encoded stack of nodes.
+     * @param root The root hash of the trie.
+     * @return The boolean validity of the proof.
+     */
+    function verify(
+        bytes32 value,
+        bytes calldata encodedPath,
+        bytes calldata rlpParentNodes,
+        bytes32 root
+    )
+        external
+        pure
+        returns (bool)
+    {
+        RLP.RLPItem memory item = RLP.toRLPItem(rlpParentNodes);
+        RLP.RLPItem[] memory parentNodes = RLP.toList(item);
+
+        bytes memory currentNode;
+        RLP.RLPItem[] memory currentNodeList;
+
+        bytes32 nodeKey = root;
+        uint pathPtr = 0;
+
+        bytes memory path = _getNibbleArray2(encodedPath);
+        if(path.length == 0) {return false;}
+
+        for (uint i = 0; i < parentNodes.length; i++) {
+            if(pathPtr > path.length) {return false;}
+
+            currentNode = RLP.toBytes(parentNodes[i]);
+            if(nodeKey != keccak256(abi.encodePacked(currentNode))) {return false;}
+            currentNodeList = RLP.toList(parentNodes[i]);
+
+            if(currentNodeList.length == 17) {
+                if(pathPtr == path.length) {
+                    if(keccak256(abi.encodePacked(RLP.toBytes(currentNodeList[16]))) == value) {
+                        return true;
+                    } else {
+                        return false;
+                    }
+                }
+
+                uint8 nextPathNibble = uint8(path[pathPtr]);
+                if(nextPathNibble > 16) {return false;}
+                nodeKey = RLP.toBytes32(currentNodeList[nextPathNibble]);
+                pathPtr += 1;
+            } else if(currentNodeList.length == 2) {
+
+                // Count of matching node key nibbles in path starting from pathPtr.
+                uint traverseLength = _nibblesToTraverse(RLP.toData(currentNodeList[0]), path, pathPtr);
+
+                if(pathPtr + traverseLength == path.length) { //leaf node
+                    if(keccak256(abi.encodePacked(RLP.toData(currentNodeList[1]))) == value) {
+                        return true;
+                    } else {
+                        return false;
+                    }
+                } else if (traverseLength == 0) { // error: couldn't traverse path
+                    return false;
+                } else { // extension node
+                    pathPtr += traverseLength;
+                    nodeKey = RLP.toBytes32(currentNodeList[1]);
+                }
+
+            } else {
+                return false;
+            }
+        }
+    }
+
+    function verifyDebug(
```

Comment:
should we clean this up, and remove here in mosaic-1? we kept it around in mosaic-0, but i think we should clean it now

Reply:
Yes, we can remove it and bring it back if needed.

Reply:
Done.